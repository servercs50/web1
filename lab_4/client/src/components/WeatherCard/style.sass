.WeatherCard

  .weather-main
    padding-bottom: 0!important

    .weather-temperature
      flex-grow: 1

  .weather-item
    padding: 0.4rem
    margin-bottom: 1rem
    background-color: #f5f5f5
    display: flex
    justify-content: space-between

    p
      margin-bottom: 0

  .weather-item:last-child
    margin-bottom: 0

.isFavorite
  .weather-items
    padding-top: 0!important
