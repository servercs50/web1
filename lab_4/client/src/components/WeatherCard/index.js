import React from 'react';

import {Grid, Paper, Fab, Hidden} from "@material-ui/core";
import Typography from '@material-ui/core/Typography/index';
import CloseIcon from '@material-ui/icons/Close';

import LoadingSpinner from '../LoadingSpinner';

import './style.sass';


export default function WeatherCard(props) {
  const data = props.cityWeatherData;;

  if (data && data.isLoading) {
    return (<LoadingSpinner/>);
  }

  if (data && data.cod && data.cod != 200) {
    if ('onFetchError' in props) props.onFetchError()
    return (
      <div>
        <p>К сожалению, не получилось получить данные о погоде...</p>
        <p>{data.message}</p>
      </div>
    )
  }

  if (data && data.cod == 200) {
    const K = 273.15; // для перевода Кельвин в Цельсия
    const temperatureCelsius = Math.round(data.main.temp - K);

    return (
      <Grid container direction={props.isFavorite ? "column" : "row"} spacing={4}
            className={`WeatherCard ${props.isFavorite ? "isFavorite" : ""}`}>

        <Grid container item xs={12} md={props.isFavorite ? 12 : 6}
              direction={props.isFavorite ? "row" : "column"} alignItems={"center"} className={"weather-main"}>
          <Grid container item sm={props.isFavorite ? 4 : false} justify={"space-between"}>
            <Typography variant="h5" component="h2">
              <b>{data.name}</b>
            </Typography>
            <Hidden smUp>
              {props.isFavorite &&
              <Fab color="primary" size={"small"}
                   onClick={props.onRemoveCityClick.bind(this, props.cityWeatherData.name)}>
                <CloseIcon/>
              </Fab>
              }
            </Hidden>
          </Grid>
          <Grid item container sm={props.isFavorite ? 7 : false} alignItems={"center"}
                justify={"space-evenly"} className={"weather-temperature"}>
            <img src={`http://openweathermap.org/img/wn/${data.weather[0].icon}@2x.png`} alt=""/>
            <Typography variant={props.isFavorite ? "h3" : "h1"} component="h2">
              {temperatureCelsius}°C
            </Typography>
          </Grid>
          <Hidden only="xs">
            <Grid container item sm={1} justify={"flex-end"}>
              {props.isFavorite &&
              <Fab color="primary" size={"small"}
                   onClick={props.onRemoveCityClick.bind(this, props.cityWeatherData.cityName)}>
                <CloseIcon/>
              </Fab>
              }
            </Grid>
          </Hidden>
        </Grid>

        <Grid item xs={12} md={props.isFavorite ? 12 : 6} className={"weather-items"}>
          <Paper elevation={1} className={"weather-item"}>
            <span><b>Ветер</b></span>
            <span>{data.wind.speed} м/с</span>
          </Paper>

          <Paper elevation={1} className={"weather-item"}>
            <span><b>Облачность</b></span>
            <span>{data.weather[0].description}</span>
          </Paper>

          <Paper elevation={1} className={"weather-item"}>
            <span><b>Давление</b></span>
            <span>{data.main.pressure} hpa</span>
          </Paper>

          <Paper elevation={1} className={"weather-item"}>
            <span><b>Влажность</b></span>
            <span>{data.main.humidity} %</span>
          </Paper>

          <Paper elevation={1} className={"weather-item"}>
            <span><b>Координаты</b></span>
            <span>[{data.coord.lon}, {data.coord.lat}]</span>
          </Paper>
        </Grid>
      </Grid>
    )
  }

  return null;
}
