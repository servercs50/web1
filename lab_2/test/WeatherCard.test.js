import React from 'react';
import {shallow} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';

import WeatherCard from '../src/components/WeatherCard';

const cityWeatherData = {
  "coord": {"lon": 37.62, "lat": 55.75},
  "weather": [{"id": 803, "main": "Clouds", "description": "broken clouds", "icon": "04n"}],
  "base": "stations",
  "main": {"temp": 283.57, "pressure": 995, "humidity": 87, "temp_min": 282.59, "temp_max": 285.15},
  "visibility": 10000,
  "wind": {"speed": 9, "deg": 230},
  "clouds": {"all": 75},
  "dt": 1572987525,
  "sys": {"type": 1, "id": 9027, "country": "RU", "sunrise": 1573015504, "sunset": 1573047671},
  "timezone": 10800,
  "id": 524901,
  "name": "Moscow",
  "cod": 200
}


describe('Weather Card', () => {

  it('renders correctly when data is loading', () => {
    const cityWeatherLoading = {
      name: 'Moscow',
      isLoading: true
    }

    const tree = shallow(
      <WeatherCard
        cityWeatherData={cityWeatherLoading}
        isFavorite
      />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly when favorite city weather data passed with cod 200', () => {
    const onClick = jest.fn()

    const tree = shallow(
      <WeatherCard
        cityWeatherData={cityWeatherData}
        isFavorite
        onRemoveCityClick={onClick}
      />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly when NON favorite city weather data passed with cod 200', () => {
    const tree = shallow(
      <WeatherCard
        cityWeatherData={cityWeatherData}
        isFavorite={false}
      />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('renders correctly when  city weather data passed with cod that not equal to 200', () => {
    const badCityWeatherData = {
      name: "Moscow",
      cod: 404
    };

    const tree = shallow(
      <WeatherCard
        cityWeatherData={badCityWeatherData}
        isFavorite
      />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('calls onFetchError when cod is not equal to 200', () => {
    const badData = {
      name: "Moscow",
      cod: 404
    }
    const onFetchError = jest.fn()

    const tree = shallow(
      <WeatherCard
        cityWeatherData={badData}
        isFavorite
        onFetchError={onFetchError}
      />);

    expect(shallowToJson(tree)).toMatchSnapshot();
    expect(onFetchError).toHaveBeenCalled()
  });


  it('returns null when no data passed', () => {
    const badData = {
      name: "Moscow"
    }
    const tree = shallow(
      <WeatherCard
        cityWeatherData={badData}
        isFavorite
      />);
    expect(shallowToJson(tree)).toEqual("");
  });

})
