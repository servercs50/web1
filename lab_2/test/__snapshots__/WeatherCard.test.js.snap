// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather Card calls onFetchError when cod is not equal to 200 1`] = `
<p>
  К сожалению, не получилось получить данные о погоде...
</p>
`;

exports[`Weather Card renders correctly when  city weather data passed with cod that not equal to 200 1`] = `
<p>
  К сожалению, не получилось получить данные о погоде...
</p>
`;

exports[`Weather Card renders correctly when NON favorite city weather data passed with cod 200 1`] = `
<WithStyles(ForwardRef(Grid))
  className="WeatherCard "
  container={true}
  direction="row"
  spacing={4}
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="weather-main"
    container={true}
    direction="column"
    item={true}
    md={6}
    xs={12}
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      item={true}
      justify="space-between"
      sm={false}
    >
      <WithStyles(ForwardRef(Typography))
        component="h2"
        variant="h5"
      >
        <b>
          Moscow
        </b>
      </WithStyles(ForwardRef(Typography))>
      <Hidden
        smUp={true}
      />
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      alignItems="center"
      className="weather-temperature"
      container={true}
      item={true}
      justify="space-evenly"
      sm={false}
    >
      <img
        alt=""
        src="http://openweathermap.org/img/wn/04n@2x.png"
      />
      <WithStyles(ForwardRef(Typography))
        component="h2"
        variant="h1"
      >
        10
        °C
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <Hidden
      only="xs"
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        item={true}
        justify="flex-end"
        sm={1}
      />
    </Hidden>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    className="weather-items"
    item={true}
    md={6}
    xs={12}
  >
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Ветер
        </b>
      </span>
      <span>
        9
         м/с
      </span>
    </WithStyles(ForwardRef(Paper))>
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Облачность
        </b>
      </span>
      <span>
        broken clouds
      </span>
    </WithStyles(ForwardRef(Paper))>
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Давление
        </b>
      </span>
      <span>
        995
         hpa
      </span>
    </WithStyles(ForwardRef(Paper))>
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Влажность
        </b>
      </span>
      <span>
        87
         %
      </span>
    </WithStyles(ForwardRef(Paper))>
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Координаты
        </b>
      </span>
      <span>
        [
        37.62
        , 
        55.75
        ]
      </span>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`Weather Card renders correctly when data is loading 1`] = `<LoadingSpinner />`;

exports[`Weather Card renders correctly when favorite city weather data passed with cod 200 1`] = `
<WithStyles(ForwardRef(Grid))
  className="WeatherCard isFavorite"
  container={true}
  direction="column"
  spacing={4}
>
  <WithStyles(ForwardRef(Grid))
    alignItems="center"
    className="weather-main"
    container={true}
    direction="row"
    item={true}
    md={12}
    xs={12}
  >
    <WithStyles(ForwardRef(Grid))
      container={true}
      item={true}
      justify="space-between"
      sm={4}
    >
      <WithStyles(ForwardRef(Typography))
        component="h2"
        variant="h5"
      >
        <b>
          Moscow
        </b>
      </WithStyles(ForwardRef(Typography))>
      <Hidden
        smUp={true}
      >
        <WithStyles(ForwardRef(Fab))
          color="primary"
          onClick={[Function]}
          size="small"
        >
          <CloseIcon />
        </WithStyles(ForwardRef(Fab))>
      </Hidden>
    </WithStyles(ForwardRef(Grid))>
    <WithStyles(ForwardRef(Grid))
      alignItems="center"
      className="weather-temperature"
      container={true}
      item={true}
      justify="space-evenly"
      sm={7}
    >
      <img
        alt=""
        src="http://openweathermap.org/img/wn/04n@2x.png"
      />
      <WithStyles(ForwardRef(Typography))
        component="h2"
        variant="h3"
      >
        10
        °C
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(Grid))>
    <Hidden
      only="xs"
    >
      <WithStyles(ForwardRef(Grid))
        container={true}
        item={true}
        justify="flex-end"
        sm={1}
      >
        <WithStyles(ForwardRef(Fab))
          color="primary"
          onClick={[Function]}
          size="small"
        >
          <CloseIcon />
        </WithStyles(ForwardRef(Fab))>
      </WithStyles(ForwardRef(Grid))>
    </Hidden>
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    className="weather-items"
    item={true}
    md={12}
    xs={12}
  >
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Ветер
        </b>
      </span>
      <span>
        9
         м/с
      </span>
    </WithStyles(ForwardRef(Paper))>
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Облачность
        </b>
      </span>
      <span>
        broken clouds
      </span>
    </WithStyles(ForwardRef(Paper))>
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Давление
        </b>
      </span>
      <span>
        995
         hpa
      </span>
    </WithStyles(ForwardRef(Paper))>
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Влажность
        </b>
      </span>
      <span>
        87
         %
      </span>
    </WithStyles(ForwardRef(Paper))>
    <WithStyles(ForwardRef(Paper))
      className="weather-item"
      elevation={1}
    >
      <span>
        <b>
          Координаты
        </b>
      </span>
      <span>
        [
        37.62
        , 
        55.75
        ]
      </span>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
